### PR TITLE
Retry EXCEEDED_MEMORY_LIMIT failures with higher memory settings

### PR DESCRIPTION
**Pull request by pgupta2**
PoS supports retring queries with different settings based on
error code. Whenever a query fails with EXCEEDED_MEMORY_LIMIT error,
we can simply increase the presto/spark memory limits and resubmit
the query again. This should resolve the OOM problems in most of the
cases.

Test plan:
* Unit Test
* Tested production workload and verified that retry logic is working correctly.

```
== RELEASE NOTES ==

Spark Changes
* Add a new configuration property ``spark.retry-on-out-of-memory-with-increased-memory-settings-enabled`` to enable picking up presto and spark memory settings and retry the query within the same spark session with the new settings applied.  This can be overridden by ``spark_retry_on_out_of_memory_with_increased_memory_settings_enabled`` session property. 
* Add new configuration properties ``spark.retry-presto-session-properties`` and ``spark.retry-spark-configs`` to alter the Presto session properties and Spark settings during retry respectively. They can be overridden by ``out_of_memory_retry_presto_session_properties`` and ``out_of_memory_retry_spark_configs`` session properties. 
```

### CODE REVIEW

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java [hunk 1]
```diff
@@ -42,6 +42,8 @@
     private double memoryRevokingThreshold;
     private double memoryRevokingTarget;
     private boolean retryOnOutOfMemoryBroadcastJoinEnabled;
+    private boolean retryOnOutOfMemoryWithIncreasedMemorySettingsEnabled;
+    private String sparkRetryIncreasedMemorySettingsJson = "{\"sparkSettings\":{},\"prestoSettings\":{}}";
```
Comment (highker) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java:46]: Will we have dynamic configs during retry? Like the oom retry, can we have the fixed override in `getRetryExecutionSettings` to avoid string encoding here?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkFailureUtils.java [hunk 2]
```diff
@@ -27,13 +28,18 @@
...
 import static com.facebook.presto.spark.classloader_interface.RetryExecutionStrategy.DISABLE_BROADCAST_JOIN;
+import static com.facebook.presto.spark.classloader_interface.RetryExecutionStrategy.INCREASE_CONTAINER_SIZE;
 import static com.facebook.presto.spi.StandardErrorCode.EXCEEDED_LOCAL_BROADCAST_JOIN_MEMORY_LIMIT;
+import static com.facebook.presto.spi.StandardErrorCode.EXCEEDED_LOCAL_MEMORY_LIMIT;
 import static com.google.common.base.Preconditions.checkState;
 import static java.util.Objects.requireNonNull;
 
 public class PrestoSparkFailureUtils
 {
+    private static final Logger log = Logger.get(PrestoSparkFailureUtils.class);
```
Comment (highker) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkFailureUtils.java:41]: Not used

File: presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkRetryExecutionUtils.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        log.info("Applying retryExecutionStrategy: " + retryExecutionStrategy.name());
+        switch (retryExecutionStrategy) {
+            case DISABLE_BROADCAST_JOIN:
+                ImmutableMap.Builder<String, String> prestoSettings = new ImmutableMap.Builder<>();
+                prestoSettings.put(JOIN_DISTRIBUTION_TYPE, FeaturesConfig.JoinDistributionType.PARTITIONED.name());
+                return new PrestoSparkRetryExecutionSettings(ImmutableMap.of(), prestoSettings.build());
+
+            case INCREASE_CONTAINER_SIZE:
+                return getSparkRetryIncreasedMemorySettings(session);
```
Comment (highker) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkRetryExecutionUtils.java:47]: Same comments as above. If we know the fixed setting here, we can avoid all the json encoding and tunneling in the code.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkRetryExecutionUtils.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+{
+    private static final Logger log = Logger.get(PrestoSparkSessionContext.class);
+
+    private PrestoSparkRetryExecutionUtils() {}
+
+    public static PrestoSparkRetryExecutionSettings getRetryExecutionSettings(
+            RetryExecutionStrategy retryExecutionStrategy,
+            Session session)
+    {
+        log.info("Applying retryExecutionStrategy: " + retryExecutionStrategy.name());
```
Comment (highker) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkRetryExecutionUtils.java:39]: nit: we can spell out "retry execution strategy: " in plain English here. Same for the one below.

**Review by highker — COMMENTED** [review]
one main comment on the json encoding and tunneling.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java (hunk 1 above)
Comment (pgupta2) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java:46]: Yes. We can have dynamic configs that can be applied on retry based on what was the original configs that were used.   My initial design was to use static set of configs but then I realized that extending that would always require a code change and release. Hence, decided to use the json encoding here. This will enable us to have different set of configs and values for different kinds of containers deployment and it will be easily extensible.

Comment (highker) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java:46]: Can we use `MapSplitter` to simplify the settings? One example is available at `TransactionManagerConfig`. We can separate the presto and spark settings into two configs:
- retryPrestoSessionProperties
- retrySparkSettings (or whatever name is used in spark).

In such case, we can avoid a lot of json codec tunneling

File: presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkRetryExecutionUtils.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+{
+    private static final Logger log = Logger.get(PrestoSparkSessionContext.class);
+
+    private PrestoSparkRetryExecutionUtils() {}
+
+    public static PrestoSparkRetryExecutionSettings getRetryExecutionSettings(
+            RetryExecutionStrategy retryExecutionStrategy,
+            Session session)
+    {
+        log.info("Applying retry execution strategy: " + retryExecutionStrategy.name());
```
Comment (highker) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkRetryExecutionUtils.java:39]: The other nit: you can also log the query id in case you don't know how to correlate these two...

**Review by highker — COMMENTED** [review]
nits only; otherwise LGTM

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java (hunk 1 above)
Comment (pgupta2) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java:46]: So, basically you are saying that instead of using one config for passing both presto and spark settings, we should add two separate configs: one for passing presto session properties and other for passing spark configs. Sounds good to me. Will do the changes.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkSessionProperties.java [hunk 6]
```diff
@@ -42,11 +44,15 @@
...
+    public static final String SPARK_RETRY_ON_OUT_OF_MEMORY_WITH_INCREASED_MEMORY_SETTINGS_ENABLED = "spark_retry_on_out_of_memory_with_increased_memory_settings_enabled";
+    public static final String OUT_OF_MEMORY_RETRY_PRESTO_SESSION_PROPERTIES = "out_of_memory_retry_presto_session_properties";
+    public static final String OUT_OF_MEMORY_RETRY_SPARK_CONFIGS = "out_of_memory_retry_spark_configs";
 
     private final List<PropertyMetadata<?>> sessionProperties;
 
     @Inject
-    public PrestoSparkSessionProperties(PrestoSparkConfig prestoSparkConfig)
+    public PrestoSparkSessionProperties(
+            PrestoSparkConfig prestoSparkConfig)
```
Comment (highker) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkSessionProperties.java:55]: We can keep the old same one line unchanged.

**Review by highker — APPROVED** [review]

**Comment by highker** [comment]
The test failure is related. Please take a look
